File: apps/gateway/src/tmux/local-shell-path.ts
import { existsSync } from 'node:fs';
import { delimiter, join } from 'node:path';

const SHELL_ENV_BEGIN_MARKER = '__TMEX_SHELL_ENV_BEGIN__';
const SHELL_ENV_END_MARKER = '__TMEX_SHELL_ENV_END__';
const SHELL_ENV_PROBE_COMMAND = `printf '${SHELL_ENV_BEGIN_MARKER}\\n'; /usr/bin/env; printf '${SHELL_ENV_END_MARKER}\\n'`;

interface RunSyncResult {
  exitCode: number;
  stdout: string;
  stderr: string;
}

interface LocalShellPathCacheDeps {
  env: NodeJS.ProcessEnv;
  fileExists: (path: string) => boolean;
  platform: NodeJS.Platform;
  runSync: (cmd: string[]) => RunSyncResult;
}

export interface LocalShellPathCache {
  get(): string | null;
  prime(): string | null;
}

function defaultRunSync(cmd: string[]): RunSyncResult {
  const result = Bun.spawnSync(cmd, {
    env: process.env,
    stdout: 'pipe',
    stderr: 'pipe',
  });

  return {
    exitCode: result.exitCode,
    stdout: Buffer.from(result.stdout).toString('utf8'),
    stderr: Buffer.from(result.stderr).toString('utf8'),
  };
}

function resolveShellFromDscl(deps: LocalShellPathCacheDeps): string | null {
  if (deps.platform !== 'darwin') {
    return null;
  }

  const username = deps.env.USER?.trim() || deps.env.LOGNAME?.trim();
  if (!username) {
    return null;
  }

  const result = deps.runSync(['/usr/bin/dscl', '.', '-read', `/Users/${username}`, 'UserShell']);
  if (result.exitCode !== 0) {
    return null;
  }

  const matched = result.stdout.match(/UserShell:\s*(\S+)/);
  const shellPath = matched?.[1]?.trim();
  if (!shellPath || !deps.fileExists(shellPath)) {
    return null;
  }

  return shellPath;
}

function resolveDefaultShell(deps: LocalShellPathCacheDeps): string | null {
  const envShell = deps.env.SHELL?.trim();
  if (envShell && deps.fileExists(envShell)) {
    return envShell;
  }

  const dsclShell = resolveShellFromDscl(deps);
  if (dsclShell) {
    return dsclShell;
  }

  const fallbackShell = deps.platform === 'darwin' ? '/bin/zsh' : '/bin/bash';
  if (deps.fileExists(fallbackShell)) {
    return fallbackShell;
  }

  return null;
}

function extractPathFromShellEnv(stdout: string): string | null {
  const beginIndex = stdout.lastIndexOf(SHELL_ENV_BEGIN_MARKER);
  if (beginIndex < 0) {
    return null;
  }

  const endIndex = stdout.indexOf(SHELL_ENV_END_MARKER, beginIndex + SHELL_ENV_BEGIN_MARKER.length);
  if (endIndex < 0) {
    return null;
  }

  const body = stdout.slice(beginIndex + SHELL_ENV_BEGIN_MARKER.length, endIndex);
  for (const rawLine of body.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line.startsWith('PATH=')) {
      continue;
    }

    const value = line.slice('PATH='.length).trim();
    return value.length > 0 ? value : null;
  }

  return null;
}

function canResolveExecutableFromPath(
  resolvedPath: string,
  executableName: string,
  deps: LocalShellPathCacheDeps
): boolean {
  for (const rawDir of resolvedPath.split(delimiter)) {
    const dir = rawDir.trim();
    if (!dir) {
      continue;
    }

    if (deps.fileExists(join(dir, executableName))) {
      return true;
    }
  }

  return false;
}

function probeShellPath(shellPath: string, deps: LocalShellPathCacheDeps): string | null {
  const attempts = [
    [shellPath, '-l', '-c', SHELL_ENV_PROBE_COMMAND],
    [shellPath, '-l', '-i', '-c', SHELL_ENV_PROBE_COMMAND],
    [shellPath, '-c', SHELL_ENV_PROBE_COMMAND],
  ];
  let fallbackPath: string | null = null;

  for (const cmd of attempts) {
    const result = deps.runSync(cmd);
    if (result.exitCode !== 0) {
      continue;
    }

    const resolvedPath = extractPathFromShellEnv(result.stdout);
    if (!resolvedPath) {
      continue;
    }

    fallbackPath ??= resolvedPath;
    if (canResolveExecutableFromPath(resolvedPath, 'tmux', deps)) {
      return resolvedPath;
    }
  }

  return fallbackPath;
}

export function createLocalShellPathCache(
  input: Partial<LocalShellPathCacheDeps> = {}
): LocalShellPathCache {
  const deps: LocalShellPathCacheDeps = {
    env: input.env ?? process.env,
    fileExists: input.fileExists ?? existsSync,
    platform: input.platform ?? process.platform,
    runSync: input.runSync ?? defaultRunSync,
  };

  let initialized = false;
  let cachedPath: string | null = null;

  return {
    get() {
      return initialized ? cachedPath : null;
    },
    prime() {
      if (initialized) {
        return cachedPath;
      }

      initialized = true;
      const shellPath = resolveDefaultShell(deps);
      if (!shellPath) {
        return null;
      }

      cachedPath = probeShellPath(shellPath, deps);
      return cachedPath;
    },
  };
}

const defaultLocalShellPathCache = createLocalShellPathCache();

export function primeLocalShellPath(): string | null {
  return defaultLocalShellPathCache.prime();
}

export function getLocalShellPath(): string | null {
  return defaultLocalShellPathCache.get();
}

// tmex 自身注入的环境变量（生产由 run.sh 经 app.env 注入 gateway 进程）。
// 这些绝不能漏进 tmex 拉起的 tmux 服务端——否则用户终端会继承：
// - 污染正常环境（NODE_ENV=production / DATABASE_URL / 各 TMEX_* 配置）；
// - 泄露密钥（TMEX_MASTER_KEY 是加密所有凭证的主密钥）。
// 绝大多数为 TMEX_ 前缀，少数非前缀键单列。
const TMEX_INJECTED_ENV_EXACT = new Set(['NODE_ENV', 'DATABASE_URL', 'GATEWAY_PORT', 'FE_PORT']);

function isTmexInjectedEnvKey(key: string): boolean {
  return key.startsWith('TMEX_') || TMEX_INJECTED_ENV_EXACT.has(key);
}

function isUtf8Locale(value: string | undefined): boolean {
  return typeof value === 'string' && /utf-?8/i.test(value);
}

export function buildLocalTmuxEnv(
  resolvedPath: string | null,
  baseEnv: NodeJS.ProcessEnv = process.env
): NodeJS.ProcessEnv {
  const nextEnv: NodeJS.ProcessEnv = {};
  for (const [key, value] of Object.entries(baseEnv)) {
    if (isTmexInjectedEnvKey(key)) {
      continue;
    }
    nextEnv[key] = value;
  }

  if (resolvedPath) {
    nextEnv.PATH = resolvedPath;
  }

  if (
    !isUtf8Locale(nextEnv.LC_ALL) &&
    (nextEnv.LC_ALL || (!isUtf8Locale(nextEnv.LC_CTYPE) && !isUtf8Locale(nextEnv.LANG)))
  ) {
    nextEnv.LC_ALL = 'C.UTF-8';
  }

  return nextEnv;
}

File: apps/gateway/src-rust/tmux/local_shell_path.rs
use std::collections::BTreeMap;
use std::env;
use std::path::Path;
use std::time::Duration;

use super::{CommandSpec, SpawnError, SpawnExecutor, SpawnPurpose};

pub const SHELL_ENV_BEGIN_MARKER: &str = "__TMEX_SHELL_ENV_BEGIN__";
pub const SHELL_ENV_END_MARKER: &str = "__TMEX_SHELL_ENV_END__";
pub const SHELL_ENV_PROBE_COMMAND: &str =
    "printf '__TMEX_SHELL_ENV_BEGIN__\\n'; /usr/bin/env; printf '__TMEX_SHELL_ENV_END__\\n'";

const ENV_PROBE_TIMEOUT: Duration = Duration::from_secs(5);
const ENV_PROBE_OUTPUT_LIMIT: usize = 1024 * 1024;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum HostPlatform {
    Unix,
    Windows,
    MacOs,
}

impl HostPlatform {
    pub fn current() -> Self {
        if cfg!(windows) {
            Self::Windows
        } else if cfg!(target_os = "macos") {
            Self::MacOs
        } else {
            Self::Unix
        }
    }
}

pub fn get_local_parking_command(platform: HostPlatform) -> &'static str {
    if platform == HostPlatform::Windows {
        "ping.exe -n 31 127.0.0.1"
    } else {
        "sleep 30"
    }
}

pub fn inherited_environment() -> BTreeMap<String, String> {
    env::vars().collect()
}

pub fn build_local_tmux_env(
    resolved_path: Option<&str>,
    base_env: &BTreeMap<String, String>,
    platform: HostPlatform,
) -> BTreeMap<String, String> {
    let case_insensitive = platform == HostPlatform::Windows;
    let mut next = base_env
        .iter()
        .filter(|(key, _)| !is_tmex_injected_env_key(key, case_insensitive))
        .map(|(key, value)| (key.clone(), value.clone()))
        .collect::<BTreeMap<_, _>>();
    if let Some(path) = resolved_path.filter(|path| !path.is_empty()) {
        set_env_value(&mut next, "PATH", path.to_owned(), case_insensitive);
    }

    let lc_all = get_env_value(&next, "LC_ALL", case_insensitive);
    let lc_ctype = get_env_value(&next, "LC_CTYPE", case_insensitive);
    let lang = get_env_value(&next, "LANG", case_insensitive);
    if !is_utf8_locale(lc_all)
        && (lc_all.is_some() || (!is_utf8_locale(lc_ctype) && !is_utf8_locale(lang)))
    {
        set_env_value(&mut next, "LC_ALL", "C.UTF-8".to_owned(), case_insensitive);
    }
    next
}

pub fn extract_path_from_shell_env(stdout: &str) -> Option<String> {
    let begin = stdout.rfind(SHELL_ENV_BEGIN_MARKER)?;
    let body_start = begin + SHELL_ENV_BEGIN_MARKER.len();
    let relative_end = stdout[body_start..].find(SHELL_ENV_END_MARKER)?;
    stdout[body_start..body_start + relative_end]
        .lines()
        .map(str::trim)
        .find_map(|line| line.strip_prefix("PATH="))
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_owned)
}

#[derive(Clone)]
pub struct LocalShellPathResolver {
    executor: SpawnExecutor,
    platform: HostPlatform,
    environment: BTreeMap<String, String>,
}

impl LocalShellPathResolver {
    pub fn new(
        executor: SpawnExecutor,
        platform: HostPlatform,
        environment: BTreeMap<String, String>,
    ) -> Self {
        Self {
            executor,
            platform,
            environment,
        }
    }

    pub async fn resolve(&self) -> Result<Option<String>, SpawnError> {
        if self.platform == HostPlatform::Windows {
            return Ok(get_env_value(&self.environment, "PATH", true)
                .map(str::trim)
                .filter(|value| !value.is_empty())
                .map(str::to_owned));
        }
        let Some(shell) = self.resolve_default_shell().await? else {
            return Ok(None);
        };
        let mut fallback = None;
        for args in [
            vec!["-l", "-c", SHELL_ENV_PROBE_COMMAND],
            vec!["-l", "-i", "-c", SHELL_ENV_PROBE_COMMAND],
            vec!["-c", SHELL_ENV_PROBE_COMMAND],
        ] {
            let output = self
                .executor
                .run_bounded(
                    CommandSpec::new(SpawnPurpose::LocalEnvironmentProbe, &shell)
                        .args(args)
                        .with_env(self.environment.clone(), true),
                    ENV_PROBE_TIMEOUT,
                    ENV_PROBE_OUTPUT_LIMIT,
                    ENV_PROBE_OUTPUT_LIMIT,
                )
                .await?;
            if output.exit_code != 0 {
                continue;
            }
            let Some(path) = extract_path_from_shell_env(&output.stdout_text()) else {
                continue;
            };
            fallback.get_or_insert_with(|| path.clone());
            if executable_in_path(&path, "tmux", self.platform) {
                return Ok(Some(path));
            }
        }
        Ok(fallback)
    }

    async fn resolve_default_shell(&self) -> Result<Option<String>, SpawnError> {
        if let Some(shell) = self
            .environment
            .get("SHELL")
            .map(String::as_str)
            .map(str::trim)
            .filter(|shell| Path::new(shell).exists())
        {
            return Ok(Some(shell.to_owned()));
        }
        if self.platform == HostPlatform::MacOs {
            let username = self
                .environment
                .get("USER")
                .or_else(|| self.environment.get("LOGNAME"))
                .map(String::as_str)
                .map(str::trim)
                .filter(|value| !value.is_empty());
            if let Some(username) = username {
                let output = self
                    .executor
                    .run_bounded(
                        CommandSpec::new(SpawnPurpose::LocalEnvironmentProbe, "/usr/bin/dscl")
                            .args([".", "-read", &format!("/Users/{username}"), "UserShell"])
                            .with_env(self.environment.clone(), true),
                        ENV_PROBE_TIMEOUT,
                        16 * 1024,
                        16 * 1024,
                    )
                    .await?;
                if output.exit_code == 0 {
                    if let Some(shell) = output
                        .stdout_text()
                        .split_once("UserShell:")
                        .map(|(_, value)| value.trim())
                        .filter(|value| Path::new(value).exists())
                    {
                        return Ok(Some(shell.to_owned()));
                    }
                }
            }
        }
        let fallback = if self.platform == HostPlatform::MacOs {
            "/bin/zsh"
        } else {
            "/bin/bash"
        };
        Ok(Path::new(fallback).exists().then(|| fallback.to_owned()))
    }
}

fn executable_in_path(path: &str, executable: &str, platform: HostPlatform) -> bool {
    let separator = if platform == HostPlatform::Windows {
        ';'
    } else {
        ':'
    };
    path.split(separator)
        .map(str::trim)
        .filter(|part| !part.is_empty())
        .any(|part| Path::new(part).join(executable).exists())
}

fn is_tmex_injected_env_key(key: &str, case_insensitive: bool) -> bool {
    let normalized = if case_insensitive {
        key.to_ascii_uppercase()
    } else {
        key.to_owned()
    };
    normalized.starts_with("TMEX_")
        || matches!(
            normalized.as_str(),
            "NODE_ENV" | "DATABASE_URL" | "GATEWAY_PORT" | "FE_PORT"
        )
}

fn get_env_value<'a>(
    environment: &'a BTreeMap<String, String>,
    key: &str,
    case_insensitive: bool,
) -> Option<&'a str> {
    environment
        .iter()
        .find(|(candidate, _)| {
            if case_insensitive {
                candidate.eq_ignore_ascii_case(key)
            } else {
                candidate.as_str() == key
            }
        })
        .map(|(_, value)| value.as_str())
}

fn set_env_value(
    environment: &mut BTreeMap<String, String>,
    key: &str,
    value: String,
    case_insensitive: bool,
) {
    let existing = environment
        .keys()
        .find(|candidate| {
            case_insensitive && candidate.as_str().eq_ignore_ascii_case(key)
                || !case_insensitive && candidate.as_str() == key
        })
        .cloned();
    environment.insert(existing.unwrap_or_else(|| key.to_owned()), value);
}

fn is_utf8_locale(value: Option<&str>) -> bool {
    value.is_some_and(|value| value.to_ascii_lowercase().replace('-', "").contains("utf8"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn local_tmux_environment_does_not_leak_gateway_configuration() {
        let environment = BTreeMap::from([
            ("HOME".to_owned(), "/Users/alice".to_owned()),
            ("PATH".to_owned(), "/usr/bin:/bin".to_owned()),
            ("SSH_AUTH_SOCK".to_owned(), "/tmp/agent".to_owned()),
            ("NODE_ENV".to_owned(), "production".to_owned()),
            ("DATABASE_URL".to_owned(), "secret.db".to_owned()),
            ("TMEX_MASTER_KEY".to_owned(), "secret".to_owned()),
        ]);
        assert_eq!(
            build_local_tmux_env(
                Some("/opt/homebrew/bin:/usr/bin:/bin"),
                &environment,
                HostPlatform::MacOs,
            ),
            BTreeMap::from([
                ("HOME".to_owned(), "/Users/alice".to_owned()),
                (
                    "PATH".to_owned(),
                    "/opt/homebrew/bin:/usr/bin:/bin".to_owned(),
                ),
                ("SSH_AUTH_SOCK".to_owned(), "/tmp/agent".to_owned()),
                ("LC_ALL".to_owned(), "C.UTF-8".to_owned()),
            ])
        );
    }

    #[test]
    fn shell_output_uses_the_last_complete_marked_environment() {
        assert_eq!(
            extract_path_from_shell_env(
                "PATH=/wrong\n__TMEX_SHELL_ENV_BEGIN__\nPATH=/first\n__TMEX_SHELL_ENV_END__\nnoise\n__TMEX_SHELL_ENV_BEGIN__\nPATH=/right\n__TMEX_SHELL_ENV_END__\n"
            ),
            Some("/right".to_owned())
        );
    }
}
